uoftblueprint/sistema
# [SIS-080] Library Backend

# Description
Implementing unfavourite(), favourtie(), and getAllLessonPlanNames() endpoints. I have also added the mkdir method and more descriptive error messages in Local. I did check if the moved lesson plan's old path was deleted (it was), I just forgot to include it here.

Ticket #080
***Testing Instructions***

-> When you test, make sure you have wiped data on your emulator before each run, otherwise you are going to get some errors if you already made all the files.

**Add these functions and imports to App.js :**
First call:


      import { MAINDIRECTORY } from './src/services/constants';
      import {writeFile, makeDirectory, checkFileExists, readDirectory} from './src/services/routes/Local.js';
      import LessonPlanService from './src/services/LessonPlanService.js';
      
      async function mkDir(){
        await makeDirectory(MAINDIRECTORY + '/Default/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson1/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson2/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/lesson3/');
        await makeDirectory(MAINDIRECTORY + '/Favourited/lesson4/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson5/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson6/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson7/');
      }
      async function writeF(){
        await writeFile(MAINDIRECTORY + '/Default/lesson1/lessonplan1.txt', "hello");
        await writeFile(MAINDIRECTORY + '/Default/lesson2/lessonplan2.txt', "whatup");
        await writeFile(MAINDIRECTORY + '/Favourited/lesson3/lessonplan3.txt', "hey");
        await writeFile(MAINDIRECTORY + '/Favourited/lesson4/lessonplan4.txt', "gahhh");
        // await writeFile(MAINDIRECTORY + '/Default/lesson5/lessonplan5.txt', "gggg");
        // await writeFile(MAINDIRECTORY + '/Default/lesson6/lessonplan6.txt', "ffff");
        // await writeFile(MAINDIRECTORY + '/Default/lesson7/lessonplan7.txt', "hhhh");
      }
      
      async function checkExists(){
        await checkFileExists(MAINDIRECTORY + '/Default/');
        await checkFileExists(MAINDIRECTORY + '/Favourtied/');
        // await checkFileExists(MAINDIRECTORY + '/Default/lesson7/');
      }
     
      async function favourite(){
        await LessonPlanService.favouriteLessonPlan('lesson1');
        await LessonPlanService.unfavouriteLessonPlan('lesson4');
        // await LessonPlanService.favouriteLessonPlan('lesson6');
        // await checkFileExists(MAINDIRECTORY + '/Favourited/lesson6/');
        // await readDirectory(MAINDIRECTORY + '/Favourited/lesson6/')
      }
      
      async function listAllPlans(){
        let allplans = await LessonPlanService.getAllLessonPlanNames();
        console.log(`ALL PLANS: ${allplans}`);
      }
      
      async function calls(){
        mkDir()
        .then(() =>{
          checkExists()
          .then(() =>{
            writeF()
            .then(() => {
              favourite()
              .then(() =>
              listAllPlans())
            })
          })
        })
        .catch(err => {
          console.error(`errpr: ${err.message}`);
        });
      }


Second call (change code and apply changes to emulator):

      import { MAINDIRECTORY } from './src/services/constants';
      import {writeFile, makeDirectory, checkFileExists, readDirectory} from './src/services/routes/Local.js';
      import LessonPlanService from './src/services/LessonPlanService.js';
      
      async function mkDir(){
        // await makeDirectory(MAINDIRECTORY + '/Default/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson1/');
        // await makeDirectory(MAINDIRECTORY + '/Default/lesson2/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/lesson3/');
        // await makeDirectory(MAINDIRECTORY + '/Favourited/lesson4/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson5/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson6/');
        await makeDirectory(MAINDIRECTORY + '/Default/lesson7/');
      }
      async function writeF(){
        // await writeFile(MAINDIRECTORY + '/Default/lesson1/lessonplan1.txt', "hello");
        // await writeFile(MAINDIRECTORY + '/Default/lesson2/lessonplan2.txt', "whatup");
        // await writeFile(MAINDIRECTORY + '/Favourited/lesson3/lessonplan3.txt', "hey");
        // await writeFile(MAINDIRECTORY + '/Favourited/lesson4/lessonplan4.txt', "gahhh");
        await writeFile(MAINDIRECTORY + '/Default/lesson5/lessonplan5.txt', "gggg");
        await writeFile(MAINDIRECTORY + '/Default/lesson6/lessonplan6.txt', "ffff");
        await writeFile(MAINDIRECTORY + '/Default/lesson7/lessonplan7.txt', "hhhh");
      }
      
      async function checkExists(){
        //await checkFileExists(MAINDIRECTORY + '/Default/');
        //await checkFileExists(MAINDIRECTORY + '/Favourtied/');
        await checkFileExists(MAINDIRECTORY + '/Default/lesson7/');
      }
      
      async function favourite(){
        // await LessonPlanService.favouriteLessonPlan('lesson1');
        // await LessonPlanService.unfavouriteLessonPlan('lesson4');
        await LessonPlanService.favouriteLessonPlan('lesson6');
        await checkFileExists(MAINDIRECTORY + '/Favourited/lesson6/');
        await readDirectory(MAINDIRECTORY + '/Favourited/lesson6/')
      }
      
      async function listAllPlans(){
        let allplans = await LessonPlanService.getAllLessonPlanNames();
        console.log(`ALL PLANS: ${allplans}`);
      }
      
      async function calls(){
        mkDir()
        .then(() =>{
          checkExists()
          .then(() =>{
            writeF()
            .then(() => {
              favourite()
              .then(() =>
              listAllPlans())
            })
          })
        })
        .catch(err => {
          console.error(`error: ${err.message}`);
        });
      }


**Then in MainNavigator add:**


    calls();


**Here are the test results, I got back:**
- First call: 

![image](https://user-images.githubusercontent.com/113125436/220753546-587fb7ee-087c-446b-86af-fb2166acc4f6.png)

- Second call:

![image](https://user-images.githubusercontent.com/113125436/220753742-1d656926-8660-4305-85a5-3d47360d4d4f.png)


## Type of change

Please delete options that are not relevant.



# Checklist:

- [X] My PR title is formatted as [SIS-\<issue number\>] \<issue name\> (eg. [SIS-010] Create UI button)
- [X] I have linked my issue/ticket on the project board to this PR
- [X] I have sent my PR to #team-sistema-prs on Slack and requested two people to review
